model/shared/src/main/scala/explore/modes/SpectroscopyModesMatrix.scala

```diff
@@ -35,10 +46,141 @@ object FocalPlane {
+
+  def decodeOptionalEnum[A: Enumerated](
+    filter:   String,
+    criteria: (String, A) => Boolean
+  ): Either[DecoderError, Option[A]] =
+    if (filter.isEmpty || filter.toLowerCase === "none") none.asRight
+    else decodeEnum[A, String](filter, criteria).map(_.some)
+
+  def decodeGmosSouthFilter(filter: NonEmptyString): Either[DecoderError, Option[GmosSouthFilter]] =
+    decodeOptionalEnum[GmosSouthFilter](filter.value, (i, f) => !f.obsolete && i === f.shortName)
+
+  def decodeGmosSouthDisperser(disperser: String): Either[DecoderError, GmosSouthDisperser] =
+    decodeEnum[GmosSouthDisperser, String](disperser, (i, f) => !f.obsolete && i === f.shortName)
+
+  def decodeGmosNorthFilter(filter: NonEmptyString): Either[DecoderError, Option[GmosNorthFilter]] =
+    decodeOptionalEnum[GmosNorthFilter](filter.value, (i, f) => !f.obsolete && i === f.shortName)
+
+  def decodeGmosNorthDisperser(disperser: String): Either[DecoderError, GmosNorthDisperser] =
+    decodeEnum[GmosNorthDisperser, String](disperser, (i, f) => !f.obsolete && i === f.shortName)
+
+  def decodeF2Filter(filter: NonEmptyString): Either[DecoderError, F2Filter] =
+    decodeEnum[F2Filter, String](filter.value, (i, f) => !f.obsolete && i === f.shortName)
+
+  def decodeF2Disperser(disperser: String): Either[DecoderError, F2Disperser] =
+    decodeEnum[F2Disperser, String](disperser, _ === _.shortName)
+
+  def decodeGpiFilter(filter: NonEmptyString): Either[DecoderError, GpiFilter] =
+    decodeEnum[GpiFilter, String](filter.value, (i, f) => !f.obsolete && i === f.shortName)
+
+  def decodeGpiDisperser(disperser: String): Either[DecoderError, GpiDisperser] =
+    decodeEnum[GpiDisperser, String](disperser, _ === _.shortName)
+
+  def decodeGnirsFilter(filter: NonEmptyString): Either[DecoderError, GnirsFilter] =
+    decodeEnum[GnirsFilter, String](filter.value, _ === _.shortName)
+
+  def decodeGnirsDisperser(disperser: String): Either[DecoderError, GnirsDisperser] =
+    decodeEnum[GnirsDisperser, String](disperser, _ === _.shortName)
+
+  def decode(
+    instrument0: Instrument,
+    disperser0:  String,
+    filter0:     NonEmptyString
+  ): Either[DecoderError, InstrumentRow] =
+    instrument0 match {
+      case i @ Instrument.GmosNorth  =>
+        (decodeGmosNorthDisperser(disperser0), decodeGmosNorthFilter(filter0)).mapN { case (d, f) =>
+          new InstrumentRow {
+            val instrument = i
+            type Disperser = GmosNorthDisperser
+            val disperser = d
+            type Filter = Option[GmosNorthFilter]
+            val filter = f
+          }
+        }
+      case i @ Instrument.GmosSouth  =>
+        (decodeGmosSouthDisperser(disperser0), decodeGmosSouthFilter(filter0)).mapN { case (d, f) =>
+          new InstrumentRow {
+            val instrument = i
+            type Disperser = GmosSouthDisperser
+            val disperser = d
+            type Filter = Option[GmosSouthFilter]
+            val filter = f
```

Comment:
These `InstrumentRow`s are spectroscopy-specific?  I think imaging will allow, for example, multiple filters.